### DISCUSSION TITLE
WAVDecoder not playing or returning expected values for file when used

### DISCUSSION THREAD

**Discussion by Gamgades** [pschatzmann/arduino-audio-tools#1577] [Q&A] [answered]
Hello,

I'm playing a .wav file from an external micro sd card through an ESP32.

When using the player_sd_i2s example to play an mp3 from the micro sd, it plays the audio and returns the correct audio info on the serial monitor. From this I know that the pinouts, PCM5102, connections, and sd card are all in working order. When changing the codec from MP3DecoderHelix to WAVDecoder, changing file extension from "mp3" to "wav", and passing this to the player class, no audio plays and the codec returns this to the monitor. 

![wavoutput](https://github.com/pschatzmann/arduino-audio-tools/assets/153458300/98c4e069-c705-438b-9e2e-055e8388b0e6)

I can see that it is pulling the correct .wav file as seen on the line:

`[I] AudioSourceSD.h : 72 - Using file /ESE.wav`

However the .wav data doesn't seem to be being pulled and populated into the decoder.

Am I able to manually set the decoder settings through the WAVDecoder class? Or am I missing on the decoder setup? 

Here is the attached code:

```
#include "AudioTools.h"
#include "AudioLibs/AudioSourceSD.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioCodecs/CodecWAV.h"


const char *startFilePath="/";
const char* ext="wav";
AudioSourceSD source(startFilePath, ext);
I2SStream i2s;
// MP3DecoderHelix decoder;
WAVDecoder decoder;
AudioPlayer player(source, i2s, decoder);
float volume = 0.5;

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = i2s.defaultConfig(TX_MODE);
  cfg.pin_ws = 26;
  cfg.pin_bck = 14;
  cfg.pin_data = 27;
  i2s.begin(cfg);


  // setup player
  player.setMetadataCallback(printMetaData);
  player.begin();
  player.setVolume(volume); 


}

void loop() {
  player.copy();
}
```

**Comment by pschatzmann** [comment, marked as answer]
The WAVDecoder supports valid [PCM WAV](https://en.wikipedia.org/wiki/WAV) files
Your file does not meet this condition: it does not have a valid WAV header so most likely the data part is invalid as well.

I suggest to create it with Audacity.

**Reply by Gamgades** [reply]
That resolved the issue thank you! Strange that the header was not replicated when copied from the music library. The file is directly from a CD rip. All I did was open the .wav in audacity then create an export which then played on the sketch. Running into stuttering issues as well as outlined in [this discussion](https://github.com/pschatzmann/arduino-audio-tools/discussions/1011) which I'm working on resolving now through the sd card speed.
